### PR TITLE
Add gradients for FilledExtrapolations

### PR DESCRIPTION
**Pull request by SomTambe**
Solves #438 and maybe #434. 
```julia
julia> x = rand(RGB,3,3)
3×3 Array{RGB{Float64},2} with eltype RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> itp2 = interpolate(x, BSpline(Linear()))
3×3 interpolate(::Array{RGB{Float64},2}, BSpline(Linear())) with element type RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> etp2 = extrapolate(itp2, RGB(.0,.0,.0))
3×3 extrapolate(interpolate(::Array{RGB{Float64},2}, BSpline(Linear())), RGB{Float64}(0.0,0.0,0.0)) with element type RGB{Float64}:
 RGB{Float64}(0.30544,0.136343,0.857564)    RGB{Float64}(0.34268,0.0164714,0.261636)  RGB{Float64}(0.0954148,0.197293,0.928313)
 RGB{Float64}(0.242728,0.960734,0.249229)   RGB{Float64}(0.414405,0.771848,0.286606)  RGB{Float64}(0.23209,0.623632,0.0502413)
 RGB{Float64}(0.921937,0.0466274,0.816069)  RGB{Float64}(0.781229,0.915429,0.926183)  RGB{Float64}(0.679568,0.737393,0.702759)

julia> Interpolations.gradient(etp2,3,3)
2-element StaticArrays.SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.4474772688335722,0.11376103826212525,0.6525178769801145)
 RGB{Float64}(-0.10166125750925725,-0.17803622073108794,-0.2234233554449263)

julia> Interpolations.gradient(etp2,4,4)
2-element StaticArrays.SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.0,0.0,0.0)
 RGB{Float64}(0.0,0.0,0.0)
```

### CODE REVIEW

**Comment by SomTambe** [comment]
cc. @johnnychen94 @DhairyaLGandhi

**Review by johnnychen94 — COMMENTED** [review]
this function is better moved to filled.jl

and most importantly, tests are needed.

**Comment by mkitti** [comment]
Let me know when this is ready for review or needs a CI run

**Comment by SomTambe** [comment]
For what types of constants should I test for here? The normal `FloatType` only right?

File: src/extrapolation/filled.jl [hunk 1]
```diff
@@ -54,6 +54,16 @@ end
     return ret
 end
 
+@inline function Interpolations.gradient(etp::FilledExtrapolation{T}, x::Vararg{Number,N}) where {T,N}
+    if checkbounds(Bool, etp.itp, x...)
+        gradient(etp.itp, x...)
```
Comment (DhairyaLGandhi) [inline src/extrapolation/filled.jl:59]: `parent(etp)` instead of `etp.itp` to be symmetric with the other `gradient` function?

Comment (SomTambe) [reply src/extrapolation/filled.jl:59]: Okay that sounds fine.

**Review by johnnychen94 — COMMENTED** [review]
> For what types of constants should I test for here? The normal FloatType only right?

FloatType should be sufficient enough. But since we're targeting DiffImages, we can add Colorant types as well (and also to `Flat()`/`Line` types). More test cases are always appreciated and help catches potential issues.

Please also add `@inferred` test to make sure type stability.

**Comment by SomTambe** [comment]
I think we should do one thing - add the `FloatType` ones here, and the `Colorant` types in the DiffImages to ensure that any of these aren't breaking for our use. What do you think @johnnychen94 ?

**Comment by johnnychen94** [comment]
> I think we should do one thing - add the FloatType ones here, and the Colorant types in the DiffImages to ensure that any of these aren't breaking for our use. What do you think @johnnychen94 ?

You won't test `Interpolations.gradient` in DIffImages, will you? The purpose of unit test is to break large things down into small parts, and test their basic functionality; only testing if the network forward/backward runs "normally" in its best scenario only makes sure things run correctly as an entire project. It doesn't tell you whether every single part of it runs normally, and it won't tell you which part should be checked when it goes wrong. In this sense, the unit test helps for future debugging. (You can't guarantee that you're still the maintainer three years later, and unit test helps future maintainers to understand things.)

It is often the case that people are too lazy to write tests, I understand it. And as I've said, more test cases are always appreciated and help catches potential issues.

**Comment by SomTambe** [comment]
@mkitti I've added the tests, I think we are good to go for the CI.

**Comment by mkitti** [comment]
Looks like we have some test issues based on type inference. The problem actually may not be in the new code, but older code. I'll have to take a closer look.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,3 +1,4 @@
+using Base: Float64, Float32
```
Comment (johnnychen94) [inline test/runtests.jl:1]: This is unnecessary.

File: test/gradient.jl [hunk 3]
```diff
@@ -199,4 +199,20 @@ using Test, Interpolations, DualNumbers, LinearAlgebra
...
     end
+
+    @testset "Extrapolations" begin
+        # out of bounds extrapolation check where fillvalue=constant
+        dtypes = [Float32, Float64]
+        dims = [1, 2, 3]
+        for dt in dtypes
+            for dm in dims
+                itp = interpolate(rand(dt, ntuple(_ -> 5, dm)...), BSpline(Linear()))
+                etp = extrapolate(itp, rand(dt))
```
Comment (johnnychen94) [inline test/gradient.jl:210]: Let's also test in-range gradients are computed correctly.

```julia
@test @inferred(Interpolations.gradient(etp, 3, 3)) == @inferred(Interpolations.gradient(itp, 3, 3))
```

File: src/extrapolation/filled.jl [hunk 4]
```diff
@@ -54,6 +54,17 @@ end
...
 
+@inline function Interpolations.gradient(etp::FilledExtrapolation{T}, x::Vararg{Number,N}) where {T,N}
+    itp = parent(etp)
+    if checkbounds(Bool, itp, x...)
+        gradient(itp, x...)
+    else
+        e = eltype(itp)
+        ptype = promote_type(e, typeof(etp.fillvalue))
+        return SVector{ndims(etp), ptype}(zeros(ptype, ndims(etp)))
+    end
```
Comment (johnnychen94) [inline src/extrapolation/filled.jl:65]: The type instability (`@inferred` failures) comes from the fact that this if-else block might return different types, hence @SomTambe you might need to double-check if `ptype` is appropriate. See https://docs.julialang.org/en/v1/manual/performance-tips/#Write-%22type-stable%22-functions for more explanation.

cc: @mkitti

**Review by johnnychen94 — COMMENTED** [review]
It's not required for the scope of this PR but it is appreciated to also add the same test cases for extrapolation with other schemes; they're currently missing as far as I can tell.

Once the test passes I believe it is good to merge.

**Comment by SomTambe** [comment]
Oh I just realized I made a really bad error while making the tests. Please wait until I update them.

File: test/gradient.jl [hunk 5]
```diff
@@ -206,12 +206,26 @@ using Test, Interpolations, DualNumbers, LinearAlgebra
...
+                else
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., 6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., 6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., -6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., -6., 6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., 6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., 6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, 6., -6., -6.)))
+                    @test all(iszero, @inferred(Interpolations.gradient(etp_n, -6., -6., -6.)))
+                end
```
Comment (johnnychen94) [inline test/gradient.jl:228]: These two seem sufficient to me.

```julia
for x in [ntuple(_->0, dm), ntuple(_->6, dm)]
```

Comment (SomTambe) [reply test/gradient.jl:228]: But this will fall short for the 2-dim and 3-dim cases where we have multiple quadrants right? It's better to get it checked for every quadrant right? That is why I hardcoded them.

Comment (johnnychen94) [reply test/gradient.jl:228]: Yeah I get your points but it's not really necessary. We only need to choose _some_ out of range points and test it. This is a more programmable way and is more friendly for future maintainers.

Comment (SomTambe) [reply test/gradient.jl:228]: Ok I have changed it.

File: src/extrapolation/filled.jl [hunk 6]
```diff
@@ -60,7 +60,7 @@ end
         gradient(itp, x...)
     else
         e = eltype(itp)
-        ptype = promote_type(e, typeof(etp.fillvalue))
+        ptype = promote_type(e, mapreduce(typeof, promote_type, x))
```
Comment (DhairyaLGandhi) [inline src/extrapolation/filled.jl:63]: Can't we let Julia's type system figure out the types? Arebts things like `itp` already parameterized?

Comment (johnnychen94) [reply src/extrapolation/filled.jl:63]: Not really, this is because interpolations gradient use `x`(the indices) to generate `δx` and thus the gradient output, to keep type stability we need to keep consistent to interpolation gradient internal.

File: src/extrapolation/filled.jl [hunk 7]
```diff
@@ -54,6 +54,17 @@ end
...
 end
 
+@inline function Interpolations.gradient(etp::FilledExtrapolation{T}, x::Vararg{Number,N}) where {T,N}
+    itp = parent(etp)
+    if checkbounds(Bool, itp, x...)
+        gradient(itp, x...)
+    else
+        e = eltype(itp)
+        ptype = promote_type(e, mapreduce(typeof, promote_type, x))
+        return SVector{ndims(etp), ptype}(zeros(ptype, ndims(etp)))
```
Comment (johnnychen94) [inline src/extrapolation/filled.jl:64]: `zeros(ptype, ndims(etp))` is unnecessary intermediate memory allocation, let's use tuple here.

```suggestion
        ptype = promote_type(eltype(itp), map(typeof, x)...)
        return SVector{ndims(etp), ptype}(ntuple(_->zero(ptype), ndims(etp)))
```

```julia
julia> etp = extrapolate(interpolate([1, 2, 3], BSpline(Linear())), 0);

# before
julia> @btime Interpolations.gradient($etp, -0.5f0);
  35.995 ns (1 allocation: 96 bytes)
# after
julia> @btime Interpolations.gradient($etp, -0.5f0);
  0.038 ns (0 allocations: 0 bytes)
```

**Comment by SomTambe** [comment]
@mkitti Could you run CI on this? I have corrected type instability and also added more tests for the other boundary conditions as there weren't any written for them.

**Review by johnnychen94 — APPROVED** [review]

**Comment by mkitti** [comment]
Looks good. Merging. Thank you.

**Comment by SomTambe** [comment]
The code is breaking for these types - 
```julia
julia> x = rand(RGB,3,3);

julia> itp2 = interpolate(x, BSpline(Linear()));

julia> etp2 = extrapolate(itp2, RGB(.0,.0,.0));

julia> Interpolations.gradient(etp2,3,3)
2-element SVector{2, RGB{Float64}} with indices SOneTo(2):
 RGB{Float64}(0.43457141859445714,0.1577128433792203,0.02876792599552047)
 RGB{Float64}(0.19412021621096764,-0.5215101620517046,0.36728033563097706)

julia> Interpolations.gradient(etp2,4,4) # wrong, should be of types RGB{Float64}
2-element SVector{2, Any} with indices SOneTo(2):
 0.0
 0.0
```
This is coming from the lines [62](https://github.com/JuliaMath/Interpolations.jl/blob/master/src/extrapolation/filled.jl#L62) in `Filled.jl` - 
```julia
...
ptype = promote_type(eltype(itp), map(typeof, x)...)
...
```
This renders `ptype` to be `Any` instead of being `RGB{Float64}`, which is further causing problems in gradients to propogate in DiffImages.jl. 
I wonder how I missed this.
cc. @johnnychen94 @DhairyaLGandhi

**Comment by johnnychen94** [comment]
> FloatType should be sufficient enough. But since we're targeting DiffImages, we can add Colorant types as well (and also to Flat()/Line types). More test cases are always appreciated and help catches potential issues.

A typical case that more test matters.

This is because Numbers can't be promoted to Colorants unless specified:

```julia
julia> promote_type(RGB{Float32}, Int)
Any # We would expect this to be RGB{Float32}
```

The reason we use `promote_type` is to mimic the output type of `gradient` for *Interpolation for type stability. Since this trick doesn't work, we might need to look deeper into the codebase to see how we can get that.

> Can't we let Julia's type system figure out the types?

Back to the comment, @DhairyaLGandhi do you mean to use `Base._return_type` to get the inferred output type?

**Comment by DhairyaLGandhi** [comment]
> did you mean to use Base._return_type to get the inferred output type?

Yes, or change the method signature to use `Vararg{S} where S`. Are there common cases where indices would not be the same type? I would consider that a bug if anything.
